### PR TITLE
feat: improve `inline` output

### PR DESCRIPTION
**Pull request by evilebottnawi**
It is not easy :+1:  Why remove `printLines`?
1. One place where we put `;` and `newlines`.
2. Easy reading.
3. For future: here we can add/remove unnecessary nodes(example - `parens` around arithmetic operation how it do `prettier` or better output `inline`).

After add `\n` in `php-parser` (already added, need just update) output will be more better :+1:

### CODE REVIEW

**Comment by evilebottnawi** [comment]
Very difficult PR, feel free to feedback

File: src/printer.js [hunk 1]
```diff
@@ -152,24 +129,134 @@ function isLastStatement(path) {
...
+    lineShouldHaveEndPHPTag(path)
+      ? concat([
+          isFirstNodeInParentNode(path) || !isPrevNodeInline(path)
+            ? hardline
+            : " ",
+          "?>"
+        ])
+      : "",
+    fileShouldEndWithHardline(path) ? hardline : ""
+  ]);
```
Comment (evilebottnawi) [inline src/printer.js]: Should be easy to read and maintenance :+1:

File: src/printer.js [hunk 2]
```diff
@@ -42,13 +43,17 @@ function getLast(arr) {
   return null;
 }
 
+function getNodeListProperty(node) {
```
Comment (mgrip) [inline src/printer.js]: i think @czosel ended up adding like a `getBody` function that does something similar?

File: src/printer.js [hunk 3]
```diff
@@ -59,41 +64,14 @@ function getNodeIndex(path) {
...
-  if (!n) {
-    return "";
-  } else if (typeof n === "string") {
-    return n;
-  }
-  if (n.kind === "program" && n.children.length === 0) {
-    return concat(["<?php", hardline]);
-  }
-  const printed = printNode(path, options, print);
+function lineShouldEndWithSemicolon(path) {
```
Comment (mgrip) [inline src/printer.js]: github is doing a terrible job displaying this diff ☹️

File: src/printer.js [hunk 4]
```diff
@@ -152,24 +130,138 @@ function isLastStatement(path) {
...
+    "namespace",
+    "class",
+    "interface",
+    "trait",
+    "traituse",
+    "declare"
+  ].includes(node.kind);
+}
+
+function isFirstNodeInParentProgramNode(path) {
```
Comment (mgrip) [inline src/printer.js]: would all of these functions make sense to go in the util file? the printer is getting crazy long

File: src/printer.js [hunk 5]
```diff
@@ -152,24 +130,138 @@ function isLastStatement(path) {
...
+}
+
+function isNextNodeInline(path) {
+  const parentNodeBody = getParentNodeListProperty(path);
+  const nodeIndex = getNodeIndex(path);
+  const nextNode = nodeIndex !== -1 && parentNodeBody[nodeIndex + 1];
+  return nextNode && nextNode.kind === "inline";
+}
+
+function shouldRemoveRemoveLines(path) {
```
Comment (mgrip) [inline src/printer.js]: this seems weird to me - is this fn specifically used for inline stuff only?

File: src/printer.js [hunk 6]
```diff
@@ -152,24 +130,138 @@ function isLastStatement(path) {
...
+
+function lineShouldHaveStartPHPTag(path) {
+  return isFirstNodeInParentProgramNode(path) || isPrevNodeInline(path);
+}
+
+function lineShouldHaveEndPHPTag(path) {
+  return isNextNodeInline(path);
+}
+
+function fileShouldEndWithHardline(path) {
```
Comment (mgrip) [inline src/printer.js]: is there a reason this should be separate from `lineShouldEndWithHardline`?

File: src/printer.js [hunk 7]
```diff
@@ -152,24 +130,138 @@ function isLastStatement(path) {
...
+  const node = path.getValue();
+  const isProgramNode = node.kind === "program";
+  return (
+    isProgramNode && node.children[node.children.length - 1].kind !== "inline"
+  );
+}
+
+function genericPrint(path, options, print) {
```
Comment (mgrip) [inline src/printer.js]: theres a lot going on here - a comment explaining the basic idea behind the logic would be good

File: src/printer.js (hunk 1 above)
Comment (mgrip) [reply src/printer.js]: agreed looks great 🎉

File: src/printer.js (hunk 7 above)
Comment (evilebottnawi) [reply src/printer.js]: :+1:

File: src/printer.js (hunk 6 above)
Comment (evilebottnawi) [reply src/printer.js]: @mgrip `lineShouldEndWithHardline` for statement, `fileShouldEndWithHardline` for `program` node - i.e. end of file

File: src/printer.js (hunk 3 above)
Comment (evilebottnawi) [reply src/printer.js]: @mgrip agree :disappointed:

File: src/printer.js (hunk 2 above)
Comment (evilebottnawi) [reply src/printer.js]: @mgrip we can rename to `getBody`, do it?

**Comment by evilebottnawi** [comment]
@mgrip i agree, it is hard to handle `inline`, but we should do it :star:

File: src/printer.js (hunk 4 above)
Comment (evilebottnawi) [reply src/printer.js]: :+1:

**Comment by evilebottnawi** [comment]
@mgrip basic refactor done :+1: Feel free to feedback

**Comment by mgrip** [comment]
we should probably have @czosel take a quick look as well since this is a big one but looks great to me! im good with merging 👍

**Comment by evilebottnawi** [comment]
@mgrip great! It is first part of imrove `inline`. I think i can do all pretty in `3-5` parts :smile:

**Comment by evilebottnawi** [comment]
Next PR will be less :+1:

**Pull request merged**